include/nbl/builtin/hlsl/limits.hlsl

```diff
@@ -0,0 +1,223 @@
+    // number of decimal digits that can be represented without change
+    NBL_CONSTEXPR_STATIC_INLINE int32_t digits10 = is_integer ? (is_bool ? 0 : S8 * 2 + S32 + int32_t(!is_signed) * S64) : float_decimal_digits;
+
+    // number of decimal digits necessary to differentiate all values of this type
+    NBL_CONSTEXPR_STATIC_INLINE int32_t max_digits10 = !is_integer ? float_decimal_max_digits : 0;
+
+    // the radix or integer base used by the representation of the given type
+    NBL_CONSTEXPR_STATIC_INLINE int32_t radix = 2;
+
+    // one more than the smallest negative power of the radix that is a valid normalized floating-point value
+    NBL_CONSTEXPR_STATIC_INLINE int32_t min_exponent = !is_integer ? float_min_exponent : 0;
+    // the smallest negative power of ten that is a valid normalized floating-point value
+    NBL_CONSTEXPR_STATIC_INLINE int32_t min_exponent10 = !is_integer ? float_min_decimal_exponent : 0;
+    // 	one more than the largest integer power of the radix that is a valid finite floating-point value
+    NBL_CONSTEXPR_STATIC_INLINE int32_t max_exponent = !is_integer ? float_max_exponent : 0;
+    // the largest integer power of 10 that is a valid finite floating-point value
+    NBL_CONSTEXPR_STATIC_INLINE int32_t max_exponent10 = !is_integer ? float_max_decimal_exponent : 0;
+
+    // identifies types which can cause arithmetic operations to trap
+    NBL_CONSTEXPR_STATIC_INLINE bool traps = false;
+    // identifies floating-point types that detect tinyness before rounding
+    NBL_CONSTEXPR_STATIC_INLINE bool tinyness_before = false;
+    
+    NBL_CONSTEXPR_STATIC_INLINE T min = impl::num_traits<T>::MIN;
+    NBL_CONSTEXPR_STATIC_INLINE T max = impl::num_traits<T>::MAX;
+    NBL_CONSTEXPR_STATIC_INLINE T lowest = is_integer ? min : -max;
+    NBL_CONSTEXPR_STATIC_INLINE T denorm_min = impl::num_traits<T>::DENORM_MIN;
+    NBL_CONSTEXPR_STATIC_INLINE T epsilon = is_integer ? 0 : (T(1) / T(1ull<<(float_digits-1)));
+    NBL_CONSTEXPR_STATIC_INLINE T round_error = is_same<T, bool>::value ? 0 : T(0.5);
+    NBL_CONSTEXPR_STATIC_INLINE T infinity = is_same<T, bool>::value ? 0 : T(1e+300 * 1e+300);
+    static T quiet_NaN() { return bit_cast<T>(impl::num_traits<T>::QUIET_NAN); }
+    static T signaling_NaN() { return bit_cast<T>(impl::num_traits<T>::SIGNALING_NAN); }
+};
+
+#else
+
+template<class T>
+struct numeric_limits : std::numeric_limits<T> 
+{
+    using Base = std::numeric_limits<T>;
+    NBL_CONSTEXPR_STATIC_INLINE T min = Base::min();
+    NBL_CONSTEXPR_STATIC_INLINE T max = Base::max();
+    NBL_CONSTEXPR_STATIC_INLINE T lowest = Base::lowest();
+    NBL_CONSTEXPR_STATIC_INLINE T denorm_min = Base::denorm_min();
+    NBL_CONSTEXPR_STATIC_INLINE T epsilon = Base::epsilon();
+    NBL_CONSTEXPR_STATIC_INLINE T round_error = Base::round_error();
+    NBL_CONSTEXPR_STATIC_INLINE T infinity = Base::infinity();
```

Comment:
hmm because of `--fast-math` on the CPU, maybe for both HLSL and C++ we should provide/store the `inf` and NaN (quiet/signalling) bitpatterns as `uintN_t` ?

Then this works around C++ flushing out Inf/NaN patterns to 0 behind our back, and HLSL not supporting bitcasts as constexpr functions!
(also allows us to implement `isnan` and `isinf` which are not prone to get fucked up with `--fast-math`, which we already have in `nbl::core::`)